Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift

```diff
@@ -0,0 +1,318 @@
+      return .closure(closure)
+    }
+
+    // We found a string literal; its contents might be JSON.
+    if let stringLit = syntax.as(StringLiteralExprSyntax.self) {
+      guard let text = stringLit.representedLiteralValue else {
+        return nil
+      }
+
+      return .stringLiteral(stringLit, text)
+    }
+
+    // Look further up the syntax tree.
+    if let parent = syntax.parent {
+      return preflightRefactoring(parent)
+    }
+
+    return nil
+  }
+}
+
+extension ConvertJSONToCodableStruct: SyntaxRefactoringCodeActionProvider {
+  static var title = "Create Codable structs from JSON"
+}
+
+extension ConvertJSONToCodableStruct {
+  fileprivate static let indentation = 4
+
+  private static func build(from jsonDictionary: Dictionary<String, Any>) -> DeclSyntax {
+    return
+      """
+      \(raw: self.buildStruct(from: jsonDictionary))
+      """
+  }
+
+  private static func buildStruct(
+    named name: String = "JSONValue",
+    at depth: Int = 0,
+    from jsonDictionary: Dictionary<String, Any>
+  ) -> String {
+    let members = self.buildStructMembers(at: depth + 1, from: jsonDictionary)
+    return
+      """
+      \(String(repeating: " ", count: depth * indentation))struct \(name): Codable {
+      \(members.joined(separator: "\n"))
+      \(String(repeating: " ", count: depth * indentation))}
+      """
+  }
+
+  private static func buildStructMembers(
+    at depth: Int,
+    from jsonDictionary: Dictionary<String, Any>
+  ) -> [String] {
+    var members: [String] = []
+
+    var uniquer: Set<String> = []
+    var nestedTypes: [(String, Dictionary<String, Any>)] = []
+    func addNestedType(_ name: String, object: Dictionary<String, Any>) {
+      guard uniquer.insert(name).inserted else {
+        return
+      }
+      nestedTypes.append((name, object))
+    }
+
+    for key in jsonDictionary.keys.sorted() {
+      guard let value = jsonDictionary[key] else {
+        continue
+      }
+
+      let type = self.jsonType(of: value, suggestion: key)
+      if case let .object(name) = type, let subObject = value as? Dictionary<String, Any> {
+        addNestedType(name, object: subObject)
+      } else if case let .array(innerType) = type,
+        let array = value as? [Any],
+        let (nesting, elementType) = innerType.outermostObject
+      {
+        if let subObject = array.unwrap(nesting) {
+          addNestedType(elementType, object: subObject)
+        }
+      }
+
+      let member = "\(String(repeating: " ", count: depth * indentation))var \(key): \(type.stringValue)"
+      members.append(member)
+    }
+
+    if !nestedTypes.isEmpty {
+      members.append("")
+      for (name, nestedType) in nestedTypes {
+        members.append(self.buildStruct(named: name, at: depth, from: nestedType))
+      }
+    }
+    return members
+  }
+
+  indirect enum JSONType {
+    case string
+    case double
+    case array(JSONType)
+    case null
+    case object(String)
+
+    var outermostObject: (Int, String)? {
```

Comment:
Instead of returning `unwraps` here and then calling `unwrap` with it, wouldn’t it be easier to just return the name and `JSONType.object` here?